### ISSUE TITLE
Netty went infinite loop if DNS is not responding

### ISSUE THREAD

**Issue by SK-Yang** [netty/netty#11099]
JVM: 1.8
Netty version: 4.1.48final

I was using netty to connect to APNS with HTTP/2. I noticed a case when DNS does not reply the IP of domain name(eg: api.push.apple.com = 17.x.x.x), netty will halt and behaved like encounter infinite loop.

```
[DEBUG] 2021-03-17 15:40:31.088 [nioEventLoopGroup-2-1] DnsQueryContext - [id: 0xa67b92d4] WRITE: UDP, [52749: /1.1.1.1:53], DefaultDnsQuestion(api.push.apple.com. IN AAAA)
[DEBUG] 2021-03-17 15:40:31.119 [nioEventLoopGroup-2-1] DnsNameResolver - [id: 0xa67b92d4] RECEIVED: UDP [26400: /1.1.1.1:53], DatagramDnsResponse(from: /1.1.1.1:53, to: /0:0:0:0:0:0:0:0:56448, 26400, QUERY(0), NoError(0), RD RA)
	DefaultDnsQuestion(api.push.apple.com. IN A)
	DefaultDnsRawRecord(api.push.apple.com. 278 IN CNAME 31B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 8 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 8 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 8 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 8 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 8 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 8 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 8 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 8 IN A 4B)
	DefaultDnsRawRecord(OPT flags:0 udp:1232 0B)
[DEBUG] 2021-03-17 15:40:31.119 [nioEventLoopGroup-2-1] DnsQueryContext - [id: 0xa67b92d4] WRITE: UDP, [15374: /1.1.1.1:53], DefaultDnsQuestion(api.push-apple.com.akadns.net. IN A)
[DEBUG] 2021-03-17 15:40:31.119 [nioEventLoopGroup-2-1] DnsNameResolver - [id: 0xa67b92d4] RECEIVED: UDP [52749: /1.1.1.1:53], DatagramDnsResponse(from: /1.1.1.1:53, to: /0:0:0:0:0:0:0:0:56448, 52749, QUERY(0), NoError(0), RD RA)
	DefaultDnsQuestion(api.push.apple.com. IN AAAA)
	DefaultDnsRawRecord(api.push.apple.com. 230 IN CNAME 31B)
	DefaultDnsRawRecord(akadns.net. 110 IN SOA 51B)
	DefaultDnsRawRecord(OPT flags:0 udp:1232 0B)
[DEBUG] 2021-03-17 15:40:31.119 [nioEventLoopGroup-2-1] DnsQueryContext - [id: 0xa67b92d4] WRITE: UDP, [37937: /1.1.1.1:53], DefaultDnsQuestion(api.push-apple.com.akadns.net. IN AAAA)
[DEBUG] 2021-03-17 15:40:31.119 [nioEventLoopGroup-2-1] DnsNameResolver - [id: 0xa67b92d4] RECEIVED: UDP [15374: /1.1.1.1:53], DatagramDnsResponse(from: /1.1.1.1:53, to: /0:0:0:0:0:0:0:0:56448, 15374, QUERY(0), NoError(0), RD RA)
	DefaultDnsQuestion(api.push-apple.com.akadns.net. IN A)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 11 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 11 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 11 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 11 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 11 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 11 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 11 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 11 IN A 4B)
	DefaultDnsRawRecord(OPT flags:0 udp:1232 0B)
[DEBUG] 2021-03-17 15:40:31.299 [nioEventLoopGroup-2-1] DnsNameResolver - [id: 0xa67b92d4] RECEIVED: UDP [37937: /1.1.1.1:53], DatagramDnsResponse(from: /1.1.1.1:53, to: /0:0:0:0:0:0:0:0:56448, 37937, QUERY(0), NoError(0), RD RA)
	DefaultDnsQuestion(api.push-apple.com.akadns.net. IN AAAA)
	DefaultDnsRawRecord(akadns.net. 180 IN SOA 54B)
	DefaultDnsRawRecord(OPT flags:0 udp:1232 0B)
```

No more lines after `2021-03-17 15:40:31.299`, but eclipse still thinking it's running after a min. I think the problem is because of the netty doesn't have a default socket timeout when it encountered network abnormal.

**Comment by normanmaurer** [comment]
Please upgrade the netty version... I think we fixed this

**Comment by normanmaurer** [comment]
Let me close this and please report back if you still encounter the issue after upgrading to latest version

**Comment by SK-Yang** [comment]
@normanmaurer
I tried 4.1.60final, this issue won't happen on tomcat or JBoss. Only on IBM's Websphere.

JDK: IBM Java 1.8.0_64
Server: IBM Websphere 8.5.5/9.0.0/9.0.5

**Comment by normanmaurer** [comment]
so you say it still happens on web sphere ?

**Comment by SK-Yang** [comment]
@normanmaurer
Yes, I can't see any unusual exceptions in Websphere's SystemOut.log or SystemErr.log file after set its log level to debug. So this is hard for me to ping point what is going on

**Comment by normanmaurer** [comment]
Can you take multiple Thread dumps when things happening ?

**Comment by SK-Yang** [comment]
I have managed to make it running and received push on phone after included `netty-tcnative 2.0.36.Final` and `netty-tcnative-boringssl-static 2.0.36.Final`. Seem like netty cannot work normally with IBM JDK SSL Provider. Here's log while tcnative libs not included.


```
[DEBUG] 2021-03-22 16:38:23.252 [WebContainer : 10] InternalLoggerFactory - Using SLF4J as the default logging framework
[DEBUG] 2021-03-22 16:38:23.254 [WebContainer : 10] ResourceLeakDetector - -Dio.netty.leakDetection.level: simple
[DEBUG] 2021-03-22 16:38:23.254 [WebContainer : 10] ResourceLeakDetector - -Dio.netty.leakDetection.targetRecords: 4
[DEBUG] 2021-03-22 16:38:23.281 [WebContainer : 10] PlatformDependent - Platform: Windows
[DEBUG] 2021-03-22 16:38:23.283 [WebContainer : 10] PlatformDependent0 - -Dio.netty.noUnsafe: false
[DEBUG] 2021-03-22 16:38:23.283 [WebContainer : 10] PlatformDependent0 - Java version: 8
[DEBUG] 2021-03-22 16:38:23.284 [WebContainer : 10] PlatformDependent0 - sun.misc.Unsafe.theUnsafe: available
[DEBUG] 2021-03-22 16:38:23.284 [WebContainer : 10] PlatformDependent0 - sun.misc.Unsafe.copyMemory: available
[DEBUG] 2021-03-22 16:38:23.285 [WebContainer : 10] PlatformDependent0 - java.nio.Buffer.address: available
[DEBUG] 2021-03-22 16:38:23.286 [WebContainer : 10] PlatformDependent0 - direct buffer constructor: available
[DEBUG] 2021-03-22 16:38:23.286 [WebContainer : 10] PlatformDependent0 - java.nio.Bits.unaligned: available, true
[DEBUG] 2021-03-22 16:38:23.286 [WebContainer : 10] PlatformDependent0 - jdk.internal.misc.Unsafe.allocateUninitializedArray(int): unavailable prior to Java9
[DEBUG] 2021-03-22 16:38:23.286 [WebContainer : 10] PlatformDependent0 - java.nio.DirectByteBuffer.<init>(long, int): available
[DEBUG] 2021-03-22 16:38:23.286 [WebContainer : 10] PlatformDependent - sun.misc.Unsafe: available
[DEBUG] 2021-03-22 16:38:23.287 [WebContainer : 10] PlatformDependent - maxDirectMemory: 2147483648 bytes (maybe)
[DEBUG] 2021-03-22 16:38:23.287 [WebContainer : 10] PlatformDependent - -Dio.netty.tmpdir: C:\WINDOWS\TEMP (java.io.tmpdir)
[DEBUG] 2021-03-22 16:38:23.287 [WebContainer : 10] PlatformDependent - -Dio.netty.bitMode: 64 (sun.arch.data.model)
[DEBUG] 2021-03-22 16:38:23.288 [WebContainer : 10] PlatformDependent - -Dio.netty.maxDirectMemory: 2147483648 bytes
[DEBUG] 2021-03-22 16:38:23.288 [WebContainer : 10] PlatformDependent - -Dio.netty.uninitializedArrayAllocationThreshold: -1
[DEBUG] 2021-03-22 16:38:23.289 [WebContainer : 10] CleanerJava6 - java.nio.ByteBuffer.cleaner(): available
[DEBUG] 2021-03-22 16:38:23.289 [WebContainer : 10] PlatformDependent - -Dio.netty.noPreferDirect: false
[DEBUG] 2021-03-22 16:38:23.299 [WebContainer : 10] AbstractByteBuf - -Dio.netty.buffer.checkAccessible: true
[DEBUG] 2021-03-22 16:38:23.299 [WebContainer : 10] AbstractByteBuf - -Dio.netty.buffer.checkBounds: true
[DEBUG] 2021-03-22 16:38:23.301 [WebContainer : 10] ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@4de0d982
[DEBUG] 2021-03-22 16:38:23.314 [WebContainer : 10] OpenSsl - netty-tcnative not in the classpath; OpenSslEngine will be unavailable.
[INFO ] 2021-03-22 16:38:23.314 [WebContainer : 10] ApnsClientBuilder - Native SSL provider not available; will use JDK SSL provider.
[DEBUG] 2021-03-22 16:38:23.351 [WebContainer : 10] JdkSslContext - Default protocols (JDK): [TLSv1] 
[DEBUG] 2021-03-22 16:38:23.351 [WebContainer : 10] JdkSslContext - Default cipher suites (JDK): [TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256, TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384, TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA, TLS_ECDHE_RSA_WITH_AES_256_CBC_SHA, TLS_RSA_WITH_AES_128_GCM_SHA256, TLS_RSA_WITH_AES_128_CBC_SHA, TLS_RSA_WITH_AES_256_CBC_SHA]
[DEBUG] 2021-03-22 16:38:23.373 [WebContainer : 10] MultithreadEventLoopGroup - -Dio.netty.eventLoopThreads: 24
[DEBUG] 2021-03-22 16:38:23.397 [WebContainer : 10] InternalThreadLocalMap - -Dio.netty.threadLocalMap.stringBuilder.initialSize: 1024
[DEBUG] 2021-03-22 16:38:23.397 [WebContainer : 10] InternalThreadLocalMap - -Dio.netty.threadLocalMap.stringBuilder.maxSize: 4096
[DEBUG] 2021-03-22 16:38:23.405 [WebContainer : 10] NioEventLoop - -Dio.netty.noKeySetOptimization: false
[DEBUG] 2021-03-22 16:38:23.405 [WebContainer : 10] NioEventLoop - -Dio.netty.selectorAutoRebuildThreshold: 512
[DEBUG] 2021-03-22 16:38:23.420 [WebContainer : 10] PlatformDependent - org.jctools-core.MpscChunkedArrayQueue: available
[DEBUG] 2021-03-22 16:38:23.456 [WebContainer : 10] DefaultDnsServerAddressStreamProvider - Default DNS servers: [/192.168.144.1:53, /8.8.8.8:53, /192.168.1.1:53] (sun.net.dns.ResolverConfiguration)
[DEBUG] 2021-03-22 16:38:23.476 [WebContainer : 10] NetUtil - -Djava.net.preferIPv4Stack: false
[DEBUG] 2021-03-22 16:38:23.476 [WebContainer : 10] NetUtil - -Djava.net.preferIPv6Addresses: false
[DEBUG] 2021-03-22 16:38:23.825 [WebContainer : 10] NetUtilInitializations - Loopback interface: lo (Software Loopback Interface 1, 127.0.0.1)
[DEBUG] 2021-03-22 16:38:23.826 [WebContainer : 10] NetUtil - Failed to get SOMAXCONN from sysctl and file \proc\sys\net\core\somaxconn. Default: 200
[DEBUG] 2021-03-22 16:38:24.231 [nioEventLoopGroup-2-1] DefaultChannelId - -Dio.netty.processId: 4272 (auto-detected)
[DEBUG] 2021-03-22 16:38:24.602 [nioEventLoopGroup-2-1] DefaultChannelId - -Dio.netty.machineId: d0:c6:37:ff:fe:fa:76:8e (auto-detected)
[DEBUG] 2021-03-22 16:38:24.634 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.numHeapArenas: 21
[DEBUG] 2021-03-22 16:38:24.634 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.numDirectArenas: 21
[DEBUG] 2021-03-22 16:38:24.634 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.pageSize: 8192
[DEBUG] 2021-03-22 16:38:24.634 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.maxOrder: 11
[DEBUG] 2021-03-22 16:38:24.634 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.chunkSize: 16777216
[DEBUG] 2021-03-22 16:38:24.634 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.smallCacheSize: 256
[DEBUG] 2021-03-22 16:38:24.635 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.normalCacheSize: 64
[DEBUG] 2021-03-22 16:38:24.635 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.maxCachedBufferCapacity: 32768
[DEBUG] 2021-03-22 16:38:24.635 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.cacheTrimInterval: 8192
[DEBUG] 2021-03-22 16:38:24.635 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.cacheTrimIntervalMillis: 0
[DEBUG] 2021-03-22 16:38:24.635 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.useCacheForAllThreads: true
[DEBUG] 2021-03-22 16:38:24.635 [nioEventLoopGroup-2-1] PooledByteBufAllocator - -Dio.netty.allocator.maxCachedByteBuffersPerChunk: 1023
[DEBUG] 2021-03-22 16:38:24.654 [nioEventLoopGroup-2-1] ByteBufUtil - -Dio.netty.allocator.type: pooled
[DEBUG] 2021-03-22 16:38:24.654 [nioEventLoopGroup-2-1] ByteBufUtil - -Dio.netty.threadLocalDirectBufferSize: 0
[DEBUG] 2021-03-22 16:38:24.654 [nioEventLoopGroup-2-1] ByteBufUtil - -Dio.netty.maxThreadLocalCharBufferSize: 16384
[DEBUG] 2021-03-22 16:38:24.802 [nioEventLoopGroup-2-1] ResourceLeakDetectorFactory - Loaded default ResourceLeakDetector: io.netty.util.ResourceLeakDetector@43508055
[DEBUG] 2021-03-22 16:38:24.812 [nioEventLoopGroup-2-1] DnsQueryContext - [id: 0xa185a6b9] WRITE: UDP, [27256: /192.168.144.1:53], DefaultDnsQuestion(api.push.apple.com. IN A)
[DEBUG] 2021-03-22 16:38:24.822 [nioEventLoopGroup-2-1] Recycler - -Dio.netty.recycler.maxCapacityPerThread: 4096
[DEBUG] 2021-03-22 16:38:24.823 [nioEventLoopGroup-2-1] Recycler - -Dio.netty.recycler.maxSharedCapacityFactor: 2
[DEBUG] 2021-03-22 16:38:24.823 [nioEventLoopGroup-2-1] Recycler - -Dio.netty.recycler.linkCapacity: 16
[DEBUG] 2021-03-22 16:38:24.823 [nioEventLoopGroup-2-1] Recycler - -Dio.netty.recycler.ratio: 8
[DEBUG] 2021-03-22 16:38:24.823 [nioEventLoopGroup-2-1] Recycler - -Dio.netty.recycler.delayedQueue.ratio: 8
[DEBUG] 2021-03-22 16:38:29.849 [nioEventLoopGroup-2-1] DnsQueryContext - [id: 0xa185a6b9] WRITE: UDP, [63277: /8.8.8.8:53], DefaultDnsQuestion(api.push.apple.com. IN A)
[DEBUG] 2021-03-22 16:38:29.913 [nioEventLoopGroup-2-1] DnsNameResolver - [id: 0xa185a6b9] RECEIVED: UDP [63277: /8.8.8.8:53], DatagramDnsResponse(from: /8.8.8.8:53, to: /0:0:0:0:0:0:0:0:54177, 63277, QUERY(0), NoError(0), RD RA)
	DefaultDnsQuestion(api.push.apple.com. IN A)
	DefaultDnsRawRecord(api.push.apple.com. 237 IN CNAME 31B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(OPT flags:0 udp:512 0B)
[DEBUG] 2021-03-22 16:38:29.915 [nioEventLoopGroup-2-1] DnsQueryContext - [id: 0xa185a6b9] WRITE: UDP, [59009: /192.168.144.1:53], DefaultDnsQuestion(api.push-apple.com.akadns.net. IN A)
[DEBUG] 2021-03-22 16:38:34.922 [nioEventLoopGroup-2-1] DnsQueryContext - [id: 0xa185a6b9] WRITE: UDP, [32981: /8.8.8.8:53], DefaultDnsQuestion(api.push-apple.com.akadns.net. IN A)
[DEBUG] 2021-03-22 16:38:34.988 [nioEventLoopGroup-2-1] DnsNameResolver - [id: 0xa185a6b9] RECEIVED: UDP [32981: /8.8.8.8:53], DatagramDnsResponse(from: /8.8.8.8:53, to: /0:0:0:0:0:0:0:0:54177, 32981, QUERY(0), NoError(0), RD RA)
	DefaultDnsQuestion(api.push-apple.com.akadns.net. IN A)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(api.push-apple.com.akadns.net. 29 IN A 4B)
	DefaultDnsRawRecord(OPT flags:0 udp:512 0B)
```

Another problem also surfaced after tcnative libs included, I'll open new thread for that.
